docs/adr/adr-014-bootstrap-from-previous-peers.md

```diff
@@ -214,67 +210,67 @@ Such callbacks are easily passable as options at header module construction time
 ),
-                        fx.Provide(newP2PExchange(*cfg)),
+     fx.Provide(newP2PExchange(*cfg)),
 ```
 
 (_Code Snippet 3.b: Example of required changes to celestia-node/nodebuilder/header/module.go_)
 
-As explained for _Code Snippet 1.b_ the event handler can handle how to deal with the ordered peer list, in this, we are portraying an example of situation where we are always interested in the top 10 peers if 10 or more peers are available.
-
-### Node Startup
+### Security Concerns: Long Range Attack
 
-When the node starts up, it will first check if the `peers` database exists. If it does not, the node will bootstrap from the hardcoded bootstrappers. If it does, the node will bootstrap from the peers in the database.
+With the ability to retrieve previously seen good peers, we should distinguish between those and the trusted peers for cases when we have expired local sync target (_i.e: subjective head_). On such cases, we would like to strictly sync and retrieve the subjective head from our trusted peers and no other, and for other syncing tasks, include the other stored good peers. To achieve this, we need to first include information about the expiry of the local sync target:
 
 ```diff
+
  // newP2PExchange constructs a new Exchange for headers.
- func newP2PExchange(cfg Config) func(
-        fx.Lifecycle,
-        ...
+ func newP2PExchange(
+        lc fx.Lifecycle,
++        storeChainHead *header.ExtendedHeader,
+        bpeers modp2p.Bootstrappers,
+        network modp2p.Network,
+        host host.Host,
         ...
-        []p2p.Option[p2p.ClientParameters],
-+       pstore peerstore.Peerstore,
- ) (libhead.Exchange[*header.ExtendedHeader], error) {
-        return func(
-                lc fx.Lifecycle,
-                ...
-                ...,
-                opts []p2p.Option[p2p.ClientParameters],
-        ) (libhead.Exchange[*header.ExtendedHeader], error) {
-                peers, err := cfg.trustedPeers(bpeers)
-                if err != nil {
-                        return nil, err
-                }
-+              list, err := pstore.Load()
-+               if err != nil {
-+                       return nil, err
-+               }
-+               peers := append(peers, list...)
-                ids := make([]peer.ID, len(peers))
-                for index, peer := range peers {
-                        ids[index] = peer.ID
-                        host.Peerstore().AddAddrs(peer.ID, peer.Addrs, peerstore.PermanentAddrTTL)
-                }
-                exchange, err := p2p.NewExchange[*header.ExtendedHeader](host, ids, conngater, opts...)
-                ...
-                return exchange, nil
-        }
- }
+        exchange, err := p2p.NewExchange(...,
+                p2p.WithParams(cfg.Client),
+                p2p.WithNetworkID[p2p.ClientParameters](network.String()),
+                p2p.WithChainID(network.String()),
++               p2p.WithSubjectiveInitialization(storeChainHead.IsExpired),
```

Comment:
could either be a method on the header or just an exported utility function, doesn't matter.

Reply:
I discussed with Rene that we need runtime detection for subjective initialization rather than build time, like in this case. The clean way to keep it without breaking existing abstractions of go-header is to introduce _interface-level options_ for the Head interface. That is, so Syncer can then do `exchange.Head(ctx, WithSubjectiveHead(head))`. This option would allow any `Head` implementation to access the subjective header, like `p2p.Exchange` which will decide which way to request basing on the header

Reply:
Implemented in PRs and in this ADR